internal/lefthook/run.go

```diff
@@ -10,6 +56,293 @@ func Run(opts *Options, hookName string, gitArgs []string) error {
+	if !ok {
+		log.Info(log.Brown(fmt.Sprintf("Can't find hook by name '%s'.", hookName)))
+		return errors.New("hook is not found")
+	}
+	if err := hook.Validate(); err != nil {
+		log.Info(log.Brown(fmt.Sprintf("Config error! %s", err)))
+		return err
+	}
+
+	startTime := time.Now()
+
+	if !isOutputDisabled(outputMeta, cfg.SkipOutput, envDisabledOutputs) {
+		log.Info(log.Cyan("Lefthook v" + version.Version()))
+		log.Info(log.Cyan("RUNNING HOOK:"), log.Bold(hookName))
+	}
+
+	spinner.Start()
+
+	var wg sync.WaitGroup
+	l.executeScripts(hookName, gitArgs, cfg.SourceDir, hook, &wg)
+	l.executeScripts(hookName, gitArgs, cfg.SourceDirLocal, hook, &wg)
+
+	for _, name := range hook.CommandsSorted {
+		wg.Add(1)
+
+		if hook.Parallel {
+			go l.executeCommand(name, gitArgs, hook, &wg)
+		} else {
+			l.executeCommand(name, gitArgs, hook, &wg)
+		}
+	}
+
+	wg.Wait()
+	spinner.Stop()
+
+	if !isOutputDisabled(outputSummary, cfg.SkipOutput, envDisabledOutputs) {
+		printSummary(time.Since(startTime))
+	}
+
+	if len(failList) == 0 {
+		return nil
+	}
+	return errors.New("Have failed script")
+}
+
+func (l *Lefthook) executeCommand(commandName string, gitArgs []string, hook *config.Hook, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if hook.Piped && isPipeBroken {
+		spinner.RestartWithMsg("\n", log.Bold(commandName), log.Brown("(SKIP BY BROKEN PIPE)"))
+		return
+	}
+
+	command := hook.Commands[commandName]
+
+	if command.DoSkip(l.repo.State()) {
+		spinner.RestartWithMsg(successSprint(log.Bold(commandName), log.Brown("(SKIP BY SETTINGS)")))
+		return
+	}
+	if isContainsExcludedTags(command.Tags, hook.ExcludeTags) {
+		spinner.RestartWithMsg(successSprint(log.Bold(commandName), log.Brown("(SKIP BY TAGS)")))
+		return
+	}
+
+	runner := command.GetRunner()
+
+	var files []string
+	filesCommand := hook.Files
+	if command.Files != "" {
+		filesCommand = command.Files
+	}
+
+	if strings.Contains(runner, subStagedFiles) {
+		files, _ = l.repo.StagedFiles()
+	} else if strings.Contains(runner, subFiles) || filesCommand != "" {
+		files, _ = git.FilesByCommand(filesCommand)
+	} else if strings.Contains(runner, pushFiles) {
+		files, _ = l.repo.PushFiles()
+	} else {
+		files, _ = l.repo.AllFiles()
+	}
+
+	log.Debug("\nFiles before filters: \n", files)
+
+	files = FilterGlob(files, command.Glob)
+	files = FilterExclude(files, command.Exclude)
+	files = FilterRelative(files, command.Root)
+
+	log.Debug("\nFiles after filters: \n", files)
+
+	var filesEsc []string
+	for _, fileName := range files {
+		if len(fileName) > 0 {
+			filesEsc = append(filesEsc, shellescape.Quote(fileName))
+		}
+	}
+	files = filesEsc
+
+	log.Debug("Files after escaping: \n", files)
+
+	runner = strings.Replace(runner, pushFiles, strings.Join(files, " "), -1)
+	runner = strings.Replace(runner, subStagedFiles, strings.Join(files, " "), -1)
+	runner = strings.Replace(runner, subAllFiles, strings.Join(files, " "), -1)
+	runner = strings.Replace(runner, subFiles, strings.Join(files, " "), -1)
```

Comment:
We do the same `strings.Join` multiple times. DRY please

Reply:
Also, it doesn't look right to me that we replace different subs with the same value. @mrexox please check

Reply:
BTW `strings.ReplaceAll` is a bit shorter and more readable

Reply:
We assume that we have only one template in the string, that's why it worked. According to the previous discussion, let's support multiple templates in the `run` string?